Sources/DeepLinking/Route+TrieRouter.swift

```diff
@@ -0,0 +1,228 @@
+
+            try trie.modify { node in
+
+                do {
+                    try node.add(routeComponents, handler: handler)
+                } catch Route.TrieNodeError.conflictingNodeHandler {
+                    throw TrieRouterError.duplicateRoute
+                } catch Route.TrieNodeError.duplicateParameterName(let parameterName) {
+                    throw TrieRouterError.invalidRoute(.duplicateParameterName(parameterName))
+                } catch Route.TrieNodeError.conflictingParameterName(let existing, let new) {
+                    throw TrieRouterError.conflictingRoute(.parameterComponent(existing: existing, new: new))
+                } catch Route.TrieNodeError.conflictingCatchAllComponent(let existing, let new) {
+                    throw TrieRouterError.conflictingRoute(.catchAllComponent(existing: existing, new: new))
+                } catch Route.TrieNodeError.misplacedCatchAllComponent(let catchAllName) {
+                    throw TrieRouterError.invalidRoute(.misplacedCatchAllComponent(catchAllName))
+                } catch {
+                    assertionFailure("🔥 Unexpected error when registering \(route)! Error: \(error)")
+                    throw TrieRouterError.invalidRoute(.unexpected(error))
+                }
+            }
+        }
+
+        /// Unregisters the given route from the router, and returns the associated handler if found.
+        ///
+        /// - Parameter route: The route to unregister
+        /// - Throws: A `TrieRouterError` error if the route is invalid or wasn't found.
+        @discardableResult
+        public func unregister(_ route: URL) throws -> AnyRouteHandler<T> {
+
+            let routeComponents = parseAnnotatedRoute(route)
+
+            return try trie.modify { node in
+
+                do {
+                    return try node.remove(routeComponents)
+                } catch Route.TrieNodeError.routeNotFound {
+                    throw TrieRouterError.routeNotFound
+                } catch Route.TrieNodeError.misplacedCatchAllComponent(let catchAllName) {
+                    throw TrieRouterError.invalidRoute(.misplacedCatchAllComponent(catchAllName))
+                } catch {
+                    assertionFailure("🔥 Unexpected error when unregistering \(route)! Error: \(error)")
+                    throw TrieRouterError.routeNotFound
+                }
+            }
+        }
+
+        /// Routes the given route by matching it against the current trie, and optionally notifies routing success
+        /// with a custom payload from the route handler.
+        ///
+        /// - Parameters:
+        ///   - route: The route to route.
+        ///   - handleCompletion: The closure to notify routing success with custom payload from the route handler.
+        /// - Throws: A `TrieRouterError` error if the route wasn't found.
+        public func route(_ route: URL, handleCompletion: ((T) -> Void)? = nil) throws {
```

Comment:
Is there any benefit on calling this method without specifying the `handleCompletion`?

Reply:
This `handleCompletion` is useful if you want use the router handlers to: 
- perform some asynchronous work (e.g. read from a file) and be notified of its completion.
- return some payload (e.g. some blob, a type safe representation of the route)
- both of the above 😄 

The need for this closure based API is essentially because `route` itself is synchronous, which means the handlers are given the routes to handle synchronously as well.

However, there might be use cases where you are not interested in any payload or completion signaling, and simply want to know if the route was handled or not when calling `route` (which you know immediately). And that's essentially why it's defined as an optional in the API.

Does this make it clearer?

Reply:
Yep, thanks for the explanation 😃